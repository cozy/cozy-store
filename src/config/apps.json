{
  "notRemovableApps": ["drive"],
  "notDisplayedApps": ["collect", "home", "settings", "store", "dataproxy"]
}
