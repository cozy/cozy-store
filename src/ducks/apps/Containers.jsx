import DiscoverComponent from 'ducks/apps/components/Discover'
import MyApplicationsComponent from 'ducks/apps/components/MyApplications'
import SidebarCategoriesComponent from 'ducks/apps/components/SidebarCategories'
import { connect } from 'react-redux'

import { translate } from 'cozy-ui/transpiled/react/I18n'

import { getInstalledApps, getRegistryApps } from './index'

const mapStateToProps = state => ({
  apps: getRegistryApps(state),
  installedApps: getInstalledApps(state),
  isFetching: state.apps.isFetching,
  isAppFetching: state.apps.isAppFetching,
  isInstalling: state.apps.isInstalling,
  isUninstalling: state.apps.isUninstalling,
  actionError: state.apps.actionError,
  fetchError: state.apps.fetchError
})

export const Discover = translate()(connect(mapStateToProps)(DiscoverComponent))

export const MyApplications = translate()(
  connect(mapStateToProps)(MyApplicationsComponent)
)

export const SidebarCategories = connect(
  state => ({
    apps: getRegistryApps(state),
    installedApps: getInstalledApps(state)
  }),
  null
)(SidebarCategoriesComponent)
