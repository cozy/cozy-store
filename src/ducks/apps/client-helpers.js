import { APP_TYPE } from './constants'

/**
 * Fetch at most 200 apps from the channel
 *
 * Optional filter can be given
 */
export const fetchAppsFromChannel = (client, channel, filter) => {
  let filterParam = ''
  // TODO do it with URL or queryParams
  if (filter) filterParam = `&filter[type]=${filter}`
  return client.stackClient.fetchJSON(
    'GET',
    `/registry?limit=300&versionsChannel=${channel}&latestChannelVersion=${channel}${filterParam}`
  )
}

/**
 * Fetch user apps or konnectors.
 * Adds type property in attributes from stack
 */
export const fetchUserApps = async (client, type) => {
  const route = type === APP_TYPE.KONNECTOR ? '/konnectors/' : '/apps/'
  const { data } = await client.stackClient.fetchJSON('GET', route)
  return data.map(x => {
    x.attributes.type = type
    return x
  })
}

export const fetchAppOrKonnector = (client, doctype, slug) => {
  // FIXME: hack to handle node type from stack for the konnectors
  const route =
    doctype === APP_TYPE.KONNECTOR || doctype === 'node' ? 'konnectors' : 'apps'
  return client.stackClient
    .fetchJSON('GET', `/${route}/${slug}`)
    .then(x => x.data)
}
