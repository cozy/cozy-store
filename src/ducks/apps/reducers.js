'use strict'

import { extend as extendI18n } from 'cozy-ui/react/I18n'
import { combineReducers } from 'redux'

// initial loading
export const LOADING_APP = 'LOADING_APP'
export const LOADING_APP_INTENT = 'LOADING_APP_INTENT'

export const FETCH_APPS = 'FETCH_APPS'
export const FETCH_APPS_SUCCESS = 'FETCH_APPS_SUCCESS'
export const FETCH_APPS_FAILURE = 'FETCH_APPS_FAILURE'

export const FETCH_APP = 'FETCH_APP'
export const FETCH_APP_SUCCESS = 'FETCH_APP_SUCCESS'
export const FETCH_APP_FAILURE = 'FETCH_APP_FAILURE'

// store an app state, to handle app version switch cancelling
export const RESTORE_APP = 'RESTORE_APP'
export const SAVE_APP = 'SAVE_APP'

export const FETCH_REGISTRY_APPS_SUCCESS = 'FETCH_REGISTRY_APPS_SUCCESS'

export const UNINSTALL_APP = 'UNINSTALL_APP'
export const UNINSTALL_APP_SUCCESS = 'UNINSTALL_APP_SUCCESS'
export const UNINSTALL_APP_FAILURE = 'UNINSTALL_APP_FAILURE'

export const INSTALL_APP = 'INSTALL_APP'
export const INSTALL_APP_SUCCESS = 'INSTALL_APP_SUCCESS'
export const INSTALL_APP_FAILURE = 'INSTALL_APP_FAILURE'

export function _sortAlphabetically(array, property) {
  return array.sort((a, b) => a[property] > b[property])
}

export function _isValidApp(app) {
  return !!app && !!app.slug && !!app.source && !!app.version
}

export function _consolidateApps(stateApps, newAppsInfos, lang) {
  const apps = new Map()
  stateApps.forEach(app => apps.set(app.slug, app))
  newAppsInfos.forEach(app => {
    const appFromState = apps.get(app.slug)
    // handle maintenance locales
    let appLocales = app.locales
    if (appLocales && appFromState && appFromState.locales) {
      for (let locale in appFromState.locales) {
        appLocales[locale] = Object.assign(
          {},
          appFromState.locales[locale],
          app.locales[locale]
        )
      }
    }
    if (appLocales && appLocales[lang]) {
      // access app locales from 'apps.slug.[...]'
      extendI18n({ apps: { [app.slug]: appLocales[lang] } })
    }
    if (appFromState) {
      apps.set(
        app.slug,
        Object.assign({}, appFromState, app, {
          locales: appLocales
        })
      )
    } else {
      apps.set(app.slug, app)
    }
  })
  return Array.from(apps.values()).filter(app => app)
}

export const list = (state = [], action = {}) => {
  switch (action.type) {
    case RESTORE_APP:
      if (_isValidApp(action.app)) {
        return _sortAlphabetically(
          _consolidateApps(
            // we completely replace it, so we remove the current from state
            state.filter(a => a.slug !== action.app.slug),
            [action.app],
            action.lang
          ),
          'slug'
        )
      } else {
        console.warn('Failed attempt to restore a saved app state.')
        return state
      }
    case FETCH_APP_SUCCESS:
      return _sortAlphabetically(
        _consolidateApps(state, [action.app], action.lang),
        'slug'
      )
    case FETCH_REGISTRY_APPS_SUCCESS:
    case FETCH_APPS_SUCCESS:
      return _sortAlphabetically(
        _consolidateApps(state, action.apps, action.lang),
        'slug'
      )
    case UNINSTALL_APP_SUCCESS:
      return state.map(app =>
        app.slug === action.slug ? { ...app, installed: false } : app
      )
    case INSTALL_APP_SUCCESS: {
      return _sortAlphabetically(
        state.map(app => {
          if (app.slug === action.installedApp.slug) {
            const nextApp = { ...app, ...action.installedApp, installed: true }
            // the available update is now installed
            if (nextApp.availableVersion) delete nextApp.availableVersion
            return nextApp
          }
          return app
        }),
        'slug'
      )
    }
    default:
      return state
  }
}

export const savedApp = (state = null, action = {}) => {
  switch (action.type) {
    case SAVE_APP:
      return _isValidApp(action.app) ? action.app : state
    case RESTORE_APP:
    case INSTALL_APP_SUCCESS: // after install we clean the previously saved app
      return null
    default:
      return state
  }
}

export const isFetching = (state = false, action = {}) => {
  switch (action.type) {
    case LOADING_APP:
    case FETCH_APPS:
      return true
    case FETCH_APPS_SUCCESS:
    case FETCH_APPS_FAILURE:
      return false
    default:
      return state
  }
}

export const isAppFetching = (state = false, action = {}) => {
  switch (action.type) {
    case LOADING_APP_INTENT:
    case FETCH_APP:
      return true
    case RESTORE_APP: // restoring the saved app cancels the app fetching
      return _isValidApp(action.app) ? false : state
    case FETCH_APP_SUCCESS:
    case FETCH_APP_FAILURE:
      return false
    default:
      return state
  }
}

export const isInstalling = (state = false, action = {}) => {
  switch (action.type) {
    case INSTALL_APP:
      return action.slug
    case INSTALL_APP_SUCCESS:
    case INSTALL_APP_FAILURE:
      return false
    default:
      return state
  }
}

export const isUninstalling = (state = false, action = {}) => {
  switch (action.type) {
    case UNINSTALL_APP:
      return true
    case UNINSTALL_APP_SUCCESS:
    case UNINSTALL_APP_FAILURE:
      return false
    default:
      return state
  }
}

export const actionError = (state = null, action = {}) => {
  switch (action.type) {
    case UNINSTALL_APP_FAILURE:
    case INSTALL_APP_FAILURE:
      return action.error
    case UNINSTALL_APP_SUCCESS:
    case INSTALL_APP_SUCCESS:
      return null
    default:
      return state
  }
}

export const fetchError = (state = null, action = {}) => {
  switch (action.type) {
    case FETCH_APPS_FAILURE:
    case FETCH_APP_FAILURE:
      return action.error
    case FETCH_APPS_SUCCESS:
    case FETCH_APP_SUCCESS:
      return null
    default:
      return state
  }
}

export const appsReducers = combineReducers({
  list,
  actionError,
  fetchError,
  isFetching,
  isAppFetching,
  isInstalling,
  isUninstalling,
  savedApp
})
