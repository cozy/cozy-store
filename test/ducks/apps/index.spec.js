'use strict'

document.body.innerHTML =
  '<div role="application" data-cozy=\'{"domain": "cozy.tools"}\' />'
window.location.protocol = 'http:'

import configureStore from 'redux-mock-store'
import thunk from 'redux-thunk'
import { combineReducers } from 'redux'

import {
  initApp,
  initAppIntent,
  getAppIconProps,
  getContext,
  _getRegistryAssetsLinks,
  _sanitizeCategories,
  _sanitizeManifest
} from 'ducks/apps'
import { appsReducers } from 'ducks/apps/reducers'
import _mockRegistryAppsResponse from './_mockRegistryAppsResponse'
import StackClient from 'cozy-stack-client'
import CozyClient from 'cozy-client'

const mockStore = configureStore([thunk])
const storeReducers = combineReducers({
  apps: appsReducers
})
const storeInitialeState = storeReducers({ apps: {} })

describe('Apps duck actions', () => {
  beforeEach(() => {
    jest.resetModules()
    jest.resetAllMocks()
  })

  let cozyClient
  beforeEach(() => {
    cozyClient = new CozyClient({
      uri: 'https://testcozy.mycozy.cloud',
      token: 'test-token'
    })
  })

  const rejectOrResolve = value => {
    if (value instanceof Error) {
      return Promise.reject(value)
    } else {
      return Promise.resolve(value)
    }
  }

  const setupClient = ({
    registry = [],
    apps = { data: [] },
    konnectors = { data: [] }
  }) => {
    StackClient.prototype.fetchJSON = jest.fn((method, url) => {
      if (url.match(/\/registry/)) return rejectOrResolve(registry)
      if (url.match(/\/apps/)) return rejectOrResolve(apps)
      if (url.match(/\/konnectors/)) return rejectOrResolve(konnectors)
    })
  }

  it('getAppIconProps should return correct icon props for <AppIcon /> component', () => {
    expect(getAppIconProps()).toMatchSnapshot()
  })

  it('initApp initialize realtime and fetch all apps', async () => {
    setupClient({
      registry: _mockRegistryAppsResponse
    })

    const store = mockStore(storeInitialeState)
    await store.dispatch(initApp(cozyClient, 'en'))
    const actions = store.getActions()
    expect(actions).toMatchSnapshot()
  })

  it('initAppIntent initialize realtime and fetch asked installed app', async () => {
    const testSlug = 'collect'
    setupClient({
      registry: _mockRegistryAppsResponse.data.find(a => a.slug === testSlug),
      apps: { data: [{ slug: 'collect' }] },
      konnectors: new Error('Mock error')
    })
    // we have to manually update the store state
    const storeState = {
      apps: Object.assign({}, storeInitialeState.apps, {
        isAppFetching: true
      })
    }
    const store = mockStore(storeState)
    await store.dispatch(initAppIntent(cozyClient, 'en', testSlug))
    const actions = store.getActions()
    expect(actions).toMatchSnapshot()
  })

  it('initAppIntent initialize realtime and fetch asked not installed konnector', async () => {
    const testSlug = 'konnector-bouilligue'
    setupClient({
      registry: _mockRegistryAppsResponse.data.find(a => a.slug === testSlug),
      apps: { data: [{ slug: 'collect' }] },
      konnectors: new Error('Mock error')
    })
    // we have to manually update the store state
    const storeState = {
      apps: Object.assign({}, storeInitialeState.apps, {
        isAppFetching: true
      })
    }
    const store = mockStore(storeState)
    await store.dispatch(initAppIntent(cozyClient, 'en', testSlug))
    const actions = store.getActions()
    expect(actions).toMatchSnapshot()
  })

  describe('getContext function', () => {
    beforeEach(() => {
      jest.resetModules()
      jest.resetAllMocks()
    })

    let contextResp
    beforeEach(() => {
      StackClient.prototype.fetchJSON = jest.fn((method, url) => {
        if (url === '/settings/context') return rejectOrResolve(contextResp)
      })
      getContext.clearCache()
    })

    it('should catch fetch error and not throw error', async () => {
      contextResp = new Error('A mock error')
      expect(await getContext(cozyClient)).toStrictEqual({})
      expect(StackClient.prototype.fetchJSON).toHaveBeenCalledTimes(1)
    })
    it('should return empty object if 404 and fetch only once', async () => {
      const err404 = new Error('Not found')
      err404.status = 404
      contextResp = err404
      expect(await getContext(cozyClient)).toStrictEqual({})
      expect(await getContext(cozyClient)).toStrictEqual({})
      expect(StackClient.prototype.fetchJSON).toHaveBeenCalledTimes(1)
    })
    it('should fetch correctly the context through the client and fetch only once', async () => {
      contextResp = { contextData: 'mock' }
      expect(await getContext(cozyClient)).toStrictEqual(contextResp)
      expect(await getContext(cozyClient)).toStrictEqual(contextResp)
      expect(StackClient.prototype.fetchJSON).toHaveBeenCalledTimes(1)
    })
  })
})

describe('Apps duck helpers', () => {
  let cozyClient
  beforeEach(() => {
    cozyClient = new CozyClient({
      uri: 'https://testcozy.mycozy.cloud',
      token: 'test-token'
    })
  })

  describe('_sanitizeCategories', () => {
    it('should return the list of the provided expected categories correctly', () => {
      // all of these categories is authorized by config/categories.json
      // so it won't be filtered
      const categories = [
        'cozy',
        'isp',
        'partners',
        'press',
        'shopping',
        'telecom'
      ]
      expect(_sanitizeCategories(categories)).toEqual(categories)
    })
    it('should filter unwanted categories', () => {
      // pressingggg is not allowed as category
      const categories = [
        'cozy',
        'isp',
        'partners',
        'pressingggg',
        'shopping',
        'telecom'
      ]
      expect(_sanitizeCategories(categories)).toEqual(
        categories.filter(c => c !== 'pressingggg')
      )
    })
    it('should return "others" as category if no categories provided', () => {
      expect(_sanitizeCategories([])).toEqual(['others'])
      expect(_sanitizeCategories(null)).toEqual(['others'])
    })
    it('should return "others" as category if categories after filtering is empty', () => {
      // pressingggg is not allowed as category
      expect(_sanitizeCategories(['pressingggg'])).toEqual(['others'])
    })
  })

  describe('_sanitizeManifest', () => {
    it('should handle legacy category property', () => {
      const testCategory = 'shop bling'
      expect(
        _sanitizeManifest({ slug: 'mock', category: testCategory })
      ).toStrictEqual({
        slug: 'mock',
        categories: [testCategory]
      })
    })
    it('should handle legacy name property locales ("en" only)', () => {
      const testName = 'Mock'
      expect(
        _sanitizeManifest({
          slug: 'mock',
          name: {
            en: testName
          }
        })
      ).toStrictEqual({
        slug: 'mock',
        name: testName
      })
    })
    it('should transform available_version property to availableVersion of provided', () => {
      const testVersion = '3.0.0'
      expect(
        _sanitizeManifest({
          slug: 'mock',
          available_version: testVersion
        })
      ).toStrictEqual({
        slug: 'mock',
        availableVersion: testVersion
      })
    })
    it('should remove incomplete or empty terms and keep complete terms', () => {
      const incompleteTerms = {
        id: 'mock-terms'
      }
      const completeTerms = {
        id: 'mock-terms',
        url: 'mock://terms',
        version: 'mock001'
      }
      expect(
        _sanitizeManifest({
          slug: 'mock',
          terms: {}
        })
      ).toStrictEqual({
        slug: 'mock'
      })
      expect(
        _sanitizeManifest({
          slug: 'mock',
          terms: incompleteTerms
        })
      ).toStrictEqual({
        slug: 'mock'
      })
      expect(
        _sanitizeManifest({
          slug: 'mock',
          terms: completeTerms
        })
      ).toStrictEqual({
        slug: 'mock',
        terms: completeTerms
      })
    })
    it('should remove incomplete or empty partnership and keep complete partnership', () => {
      const incompletePartnership = {
        icon: 'icon.svg' // icon is optional
      }
      const completePartnership = {
        description: 'A partnership text here' // description is mandatory
      }
      expect(
        _sanitizeManifest({
          slug: 'mock',
          partnership: {}
        })
      ).toStrictEqual({
        slug: 'mock'
      })
      expect(
        _sanitizeManifest({
          slug: 'mock',
          partnership: incompletePartnership
        })
      ).toStrictEqual({
        slug: 'mock'
      })
      expect(
        _sanitizeManifest({
          slug: 'mock',
          partnership: completePartnership
        })
      ).toStrictEqual({
        slug: 'mock',
        partnership: completePartnership
      })
    })
  })

  describe('_getRegistryAssetsLinks', () => {
    it('should handle empty manifest or appVersion', () => {
      expect(_getRegistryAssetsLinks(cozyClient, {})).toStrictEqual({})
      expect(_getRegistryAssetsLinks(cozyClient)).toStrictEqual({})
      expect(_getRegistryAssetsLinks(cozyClient, {}, '1.0.0')).toMatchSnapshot()
    })
    it('should handle icon link from manifest slug and version', () => {
      expect(
        _getRegistryAssetsLinks(cozyClient, { slug: 'mock' }, '1.0.0')
      ).toMatchSnapshot()
    })
    it('should handle screenshots links', () => {
      expect(
        _getRegistryAssetsLinks(
          cozyClient,
          {
            slug: 'mock',
            screenshots: ['screen1.jpg', 'screen2.png', '/screen3.gif']
          },
          '1.0.0'
        )
      ).toMatchSnapshot()
    })
    it('should handle partnership icon link', () => {
      expect(
        _getRegistryAssetsLinks(
          cozyClient,
          {
            slug: 'mock',
            partnership: {
              description: 'a text here',
              icon: 'partnershipIcon.svg'
            }
          },
          '1.0.0'
        )
      ).toMatchSnapshot()
    })
  })
})
