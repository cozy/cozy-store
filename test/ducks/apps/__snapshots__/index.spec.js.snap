// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Apps duck actions getAppIconProps should return correct icon props for <AppIcon /> component 1`] = `
Object {
  "domain": "cozy.tools",
  "secure": false,
}
`;

exports[`Apps duck actions initApp initialize realtime and fetch all apps 1`] = `
Array [
  Object {
    "type": "LOADING_APP",
  },
  Object {
    "type": "FETCH_APPS",
  },
  Object {
    "apps": Array [
      Object {
        "categories": Array [
          "isp",
          "telecom",
        ],
        "developer": Object {
          "name": "Cozy",
        },
        "icon": "<svg></svg>",
        "installed": false,
        "isInRegistry": true,
        "label": 2,
        "links": Object {
          "icon": "/registry/konnector-bouilligue/0.1.0/icon",
        },
        "name": "Bouilligue",
        "slug": "konnector-bouilligue",
        "type": "konnector",
        "uninstallable": true,
        "version": "0.1.0",
        "versions": Object {
          "beta": Array [
            "0.1.0",
          ],
          "dev": Array [
            "0.1.0",
          ],
          "stable": Array [
            "0.1.0",
          ],
        },
      },
      Object {
        "categories": Array [
          "transport",
        ],
        "developer": Object {
          "name": "Cozy",
        },
        "icon": "<svg></svg>",
        "installed": false,
        "isInRegistry": true,
        "label": 1,
        "links": Object {
          "icon": "/registry/konnector-trinlane/0.1.0/icon",
        },
        "name": "Trinlane",
        "permissions": Object {
          "mock": Object {
            "type": "io.mock.doctype",
          },
        },
        "slug": "konnector-trinlane",
        "tags": Array [
          "transport",
          "files",
          "bills",
        ],
        "type": "konnector",
        "uninstallable": true,
        "version": "0.1.0",
        "versions": Object {
          "beta": Array [
            "0.1.0",
          ],
          "dev": Array [
            "0.1.0",
          ],
          "stable": Array [
            "0.1.0",
          ],
        },
      },
      Object {
        "categories": Array [
          "cozy",
        ],
        "developer": Object {
          "name": "Cozy",
        },
        "editor": "Cozy",
        "icon": "<svg></svg>",
        "installed": false,
        "isInRegistry": true,
        "label": 1,
        "links": Object {
          "icon": "/registry/photos/3.0.0/icon",
        },
        "locales": Object {
          "en": Object {
            "long_description": "A long description finally short",
          },
        },
        "name": "Photos",
        "name_prefix": "Cozy",
        "slug": "photos",
        "type": "webapp",
        "uninstallable": true,
        "version": "3.0.0",
        "versions": Object {
          "beta": Array [
            "3.0.0",
          ],
          "dev": Array [
            "3.0.0",
          ],
          "stable": Array [
            "3.0.0",
          ],
        },
      },
      Object {
        "categories": Array [
          "partners",
        ],
        "editor": "vestrejail",
        "icon": "<svg></svg>",
        "installed": false,
        "isInRegistry": true,
        "label": 4,
        "links": Object {
          "icon": "/registry/tasky/1.0.0/icon",
        },
        "name": "Tasky",
        "slug": "tasky",
        "type": "webapp",
        "uninstallable": true,
        "version": "1.0.0",
        "versions": Object {
          "beta": Array [
            "1.0.0",
          ],
          "dev": Array [
            "1.0.0",
          ],
          "stable": Array [
            "1.0.0",
          ],
        },
      },
    ],
    "lang": "en",
    "type": "FETCH_REGISTRY_APPS_SUCCESS",
  },
  Object {
    "type": "FETCH_APPS",
  },
  Object {
    "apps": Array [],
    "lang": "en",
    "type": "FETCH_APPS_SUCCESS",
  },
]
`;

exports[`Apps duck actions initAppIntent initialize realtime and fetch asked installed app 1`] = `
Array [
  Object {
    "type": "LOADING_APP_INTENT",
  },
  Object {
    "type": "FETCH_APP",
  },
  Object {
    "app": Object {
      "categories": Array [
        "cozy",
      ],
      "developer": Object {
        "name": "Cozy",
      },
      "editor": "Cozy",
      "icon": "<svg></svg>",
      "installed": true,
      "isInRegistry": true,
      "label": 1,
      "links": Object {
        "icon": "/registry/collect/3.0.0/icon",
      },
      "name": "Collect",
      "name_prefix": "Cozy",
      "permissions": Object {
        "mock": Object {
          "type": "io.mock.doctype",
        },
        "mock2": Object {
          "type": "io.mock.doctype2",
        },
      },
      "slug": "collect",
      "tags": Array [
        "konnector",
        "collect",
        "bills",
        "providers",
        "files",
      ],
      "type": "webapp",
      "uninstallable": true,
      "version": "3.0.0",
      "versions": Object {
        "beta": Array [
          "3.0.0",
        ],
        "dev": Array [
          "3.0.0",
        ],
        "stable": Array [
          "3.0.0",
        ],
      },
    },
    "lang": "en",
    "type": "FETCH_APP_SUCCESS",
  },
]
`;

exports[`Apps duck actions initAppIntent initialize realtime and fetch asked not installed konnector 1`] = `
Array [
  Object {
    "type": "LOADING_APP_INTENT",
  },
  Object {
    "type": "FETCH_APP",
  },
  Object {
    "app": Object {
      "categories": Array [
        "isp",
        "telecom",
      ],
      "developer": Object {
        "name": "Cozy",
      },
      "icon": "<svg></svg>",
      "installed": false,
      "isInRegistry": true,
      "label": 2,
      "links": Object {
        "icon": "/registry/konnector-bouilligue/0.1.0/icon",
      },
      "name": "Bouilligue",
      "slug": "konnector-bouilligue",
      "type": "konnector",
      "uninstallable": true,
      "version": "0.1.0",
      "versions": Object {
        "beta": Array [
          "0.1.0",
        ],
        "dev": Array [
          "0.1.0",
        ],
        "stable": Array [
          "0.1.0",
        ],
      },
    },
    "lang": "en",
    "type": "FETCH_APP_SUCCESS",
  },
]
`;

exports[`Apps duck helpers _getRegistryAssetsLinks should handle empty manifest or appVersion 1`] = `
Object {
  "iconLink": undefined,
  "partnershipIconLink": undefined,
  "screenshotsLinks": undefined,
}
`;

exports[`Apps duck helpers _getRegistryAssetsLinks should handle icon link from manifest slug and version 1`] = `
Object {
  "iconLink": "/registry/mock/1.0.0/icon",
  "partnershipIconLink": undefined,
  "screenshotsLinks": undefined,
}
`;

exports[`Apps duck helpers _getRegistryAssetsLinks should handle partnership icon link 1`] = `
Object {
  "iconLink": "/registry/mock/1.0.0/icon",
  "partnershipIconLink": "https://testcozy.mycozy.cloud/registry/mock/1.0.0/partnership_icon",
  "screenshotsLinks": undefined,
}
`;

exports[`Apps duck helpers _getRegistryAssetsLinks should handle screenshots links 1`] = `
Object {
  "iconLink": "/registry/mock/1.0.0/icon",
  "partnershipIconLink": undefined,
  "screenshotsLinks": Array [
    "https://testcozy.mycozy.cloud/registry/mock/1.0.0/screenshots/screen1.jpg",
    "https://testcozy.mycozy.cloud/registry/mock/1.0.0/screenshots/screen2.png",
    "https://testcozy.mycozy.cloud/registry/mock/1.0.0/screenshots/screen3.gif",
  ],
}
`;
